trie-db/src/iterator.rs

```diff
@@ -0,0 +1,647 @@
+                        self.key_nibbles.append_partial(slice.right());
+
+                        let prefix = key.back(full_key_nibbles);
+                        self.db.get_raw_or_lookup(item, prefix.left())?
+                    },
+                    Node::Branch(nodes, _) => {
+                        if partial.is_empty() {
+                            return Ok(())
+                        }
+
+                        let i = partial.at(0);
+                        crumb.status = Status::AtChild(i as usize);
+                        self.key_nibbles.push(i);
+
+                        if let Some(child) = nodes[i as usize] {
+                            full_key_nibbles += 1;
+                            partial = partial.mid(1);
+
+                            let prefix = key.back(full_key_nibbles);
+                            self.db.get_raw_or_lookup(child, prefix.left())?
+                        } else {
+                            return Ok(())
+                        }
+                    },
+                    Node::NibbledBranch(slice, nodes, _) => {
+                        if !partial.starts_with(&slice) {
+                            if slice < partial {
+                                crumb.status = Status::Exiting;
+                                self.key_nibbles.append_partial(slice.right());
+                                self.key_nibbles.push((nibble_ops::NIBBLE_LENGTH - 1) as u8);
+                            }
+                            return Ok(());
+                        }
+
+                        full_key_nibbles += slice.len();
+                        partial = partial.mid(slice.len());
+
+                        if partial.is_empty() {
+                            return Ok(())
+                        }
+
+                        let i = partial.at(0);
+                        crumb.status = Status::AtChild(i as usize);
+                        self.key_nibbles.append_partial(slice.right());
+                        self.key_nibbles.push(i);
+
+                        if let Some(child) = nodes[i as usize] {
+                            full_key_nibbles += 1;
+                            partial = partial.mid(1);
+
+                            let prefix = key.back(full_key_nibbles);
+                            self.db.get_raw_or_lookup(child, prefix.left())?
+                        } else {
+                            return Ok(())
+                        }
+                    },
+                    Node::Empty => {
+                        if !partial.is_empty() {
+                            crumb.status = Status::Exiting;
+                        }
+                        return Ok(())
+                    },
+                }
+            };
+
+            node_data = next_node_data;
+            inline = next_inline;
+        }
+    }
+
+    /// Descend into a payload.
+    fn descend<'b, 'c>(&'b mut self, node_data: &'c [u8], inline: bool)
```

Comment:
I think the idea was that in the return raw case it would return a borrowed, but that's not possible due to [an issue with elastic_array](https://github.com/debris/elastic-array/issues/10), so it did a copy anyway. But in either case, it doesn't seem like a big deal to do the copy because if it returns a raw node, then it is smaller than the size of a hash anyway, so the copy is very cheap.